### PR TITLE
Generic task queue with worker pool

### PR DESCRIPTION
**Pull request by adzialocha**
Generic task queue with worker pool with a nice test simulating a jigsaw puzzle game. Closes: #83 

**Todo**

- [x] Better naming of all traits and structs
- [x] Add puzzle test
- [x] Add documentation

**Next Steps**

These issues came up during development of this PR but it makes sense to tackle them in separate PRs, together with refactoring the current `TaskManager` in `tasks.rs`:

* #87 
* #86 
* #85 

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header

### CODE REVIEW

**Comment by sandreae** [comment]
:heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: good memories :rofl:

**Comment by adzialocha** [comment]
> heart heart heart heart heart heart heart heart good memories rofl

I knoow! :heart: Of course, please commit anything here!

**Comment by sandreae** [comment]
I like the new naming :+1:

**Comment by adzialocha** [comment]
I've started the puzzle test yesterday, but I'm not sure what the workers did again ..

* "Pick worker" Pick incoming piece p and move it into database, For every related puzzle piece of p and p itself dispatch a dependency task
* "Dependency worker" Check if dependencies exist for piece g, if yes, flip "connected" flag of g to true and move them to same puzzle h, if true dispatch finish puzzle task for puzzle h
* "Finish worker" Check if all puzzle pieces in h are flipped to true, if yes, flip "complete" flag of h to true?

**Comment by sandreae** [comment]
Yes! I think that's exactly it.

We skipped any concept of "materialising a puzzle piece".

File: aquadoggo/src/worker.rs [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+
+                    pieces.push(JigsawPiece {
+                        id: (offset + id) as usize,
+                        relations,
+                    });
+                }
+            }
+
+            offset = offset + (size * size);
+        }
```
Comment (sandreae) [inline aquadoggo/src/worker.rs:712]: I can't quite grok what's happening here... can you expand a little?

Comment (sandreae) [inline aquadoggo/src/worker.rs:712]: I can't quite grok what's happening here... can you expand a little?

**Review by sandreae — APPROVED** [review]
Huhu! I liked looking through this again, so great to have a custom task queue for everything we will be working on shortly. It was also useful for the sake of rust learning ;-p thanks for tidying it up and the renaming has worked well I think. 

The docs are also niiice :+1:

There is one section of the test setup I can't get my head around, but that's a small detail I would just like to get if I can.

**Review by sandreae — APPROVED** [review]
Huhu! I liked looking through this again, so great to have a custom task queue for everything we will be working on shortly. It was also useful for the sake of rust learning ;-p thanks for tidying it up and the renaming has worked well I think. 

The docs are also niiice :+1:

There is one section of the test setup I can't get my head around, but that's a small detail I would just like to get if I can.

File: aquadoggo/src/worker.rs (hunk 1 above)
Comment (adzialocha) [reply aquadoggo/src/worker.rs]: I've added some comments, does that help?

Comment (sandreae) [reply aquadoggo/src/worker.rs]: Yes, perfect! Thanks.

**Pull request merged**